File: module/ngflink/src/main/java/me/samei/xtool/esreporter/v1/flink/reporter/JobTask.java
package me.samei.xtool.esreporter.v1.flink.reporter;

import org.apache.flink.metrics.Metric;
import org.apache.flink.metrics.MetricGroup;

import java.util.Map;

public class JobTask extends GroupsAbstract {

    private final String SYM_JOB_NAME = "<job_name>";
    private final String SYM_TASK_NAME = "<task_name>";
    private final String SYM_SUBTASK_INDEX = "<subtask_index>";

    protected String[] keys = {
            SYM_JOB_NAME,
            SYM_TASK_NAME,
            SYM_SUBTASK_INDEX
    };

    @Override
    public String select(Metric metric, String name, MetricGroup group) {
        Map<String, String> vars = group.getAllVariables();
        logger.debug("VARS: {}", vars.size());
        if (vars.size() < keys.length) return null;
        for(String key: keys) {
            logger.debug("KEY: '{}', VAR: '{}'", key, vars.get(key));
            if (!vars.containsKey(key)) return null;
            if (vars.get(key) == null) return null;
            if (vars.get(key).isEmpty()) return null;
        }
        return name;
    }

    @Override
    public String identity(Metric metric, String metricName, MetricGroup group) {
        Map<String, String> vars = group.getAllVariables();
        logger.debug("KEY: '{}', VAR: '{}'", SYM_JOB_NAME, vars.get(SYM_JOB_NAME));
        logger.debug("KEY: '{}', VAR: '{}'", SYM_TASK_NAME, vars.get(SYM_TASK_NAME));
        logger.debug("KEY: '{}', VAR: '{}'", SYM_SUBTASK_INDEX, vars.get(SYM_SUBTASK_INDEX));
        String result = vars.get(SYM_JOB_NAME) + "." + vars.get(SYM_TASK_NAME) + "." + vars.get(SYM_SUBTASK_INDEX);
        logger.debug("RESULT: {}", result);
        return result;
    }
}

File: module/flink14/src/main/scala/com/sameei/xtool/elasticreporter/v1/flink/reporter/JobTask.scala
package com.sameei.xtool.elasticreporter.v1.flink.reporter

import com.sameei.xtool.elasticreporter.v1.flink.lego
import com.sameei.xtool.elasticreporter.v1.flink.lego.ReporterForMultipleGroups
import org.apache.flink.metrics.{Metric, MetricGroup}
import scala.collection.JavaConverters._

class JobTask extends lego.ReporterForMultipleGroups {

    val JobName = "<job_name>"
    val JobId = "<job_id>"
    val TaskName = "<task_name>"
    val TaskId = "<task_id>"
    val SubtaskIndex = "<subtask_index>"

    protected val keys = Array(JobName, JobName, TaskName, TaskId, SubtaskIndex)

    protected val filterKeys = Array(
        "<host>", "<tm_id>", "<job_id>", "<job_name>",
        "<task_id>", "<task_name>", "<task_attempt_id>", "<task_attempt_num>", "<subtask_index>",
        "<operator_id>", "<operator_name>"
    )

    override protected def select(
        name : String, metric : Metric, group : MetricGroup
    ) : Option[ReporterForMultipleGroups.Ref] = {
        val vars = group.getAllVariables.asScala

        if (vars.size < keys.size) None
        else {

            val count = keys.foldLeft(0) { (count, k) =>
                if (vars.contains(k)) count + 1
                else count
            }

            if (count < keys.size) None
            else {
                val groupdId = s"${vars(JobName)}.${vars(TaskId)}.${vars(SubtaskIndex)}"
                group.getScopeComponents
                val metricKey = s"${name}."
                ???
            }
        }
    }
}
